app/web_hub.go

```diff
@@ -130,22 +135,27 @@ func (a *App) HubStop() {
 		mlog.Warn("We appear to have already sent the stop checking for deadlocks command")
 	}
 
-	for _, hub := range a.Srv().Hubs {
+	for _, hub := range a.Srv().GetHubs() {
 		hub.Stop()
 	}
 
-	a.Srv().Hubs = []*Hub{}
+	a.Srv().SetHubs([]*Hub{})
 }
 
 func (a *App) GetHubForUserId(userId string) *Hub {
-	if len(a.Srv().Hubs) == 0 {
+	if len(a.Srv().GetHubs()) == 0 {
 		return nil
 	}
 
 	hash := fnv.New32a()
 	hash.Write([]byte(userId))
-	index := hash.Sum32() % uint32(len(a.Srv().Hubs))
-	return a.Srv().Hubs[index]
+	index := hash.Sum32() % uint32(len(a.Srv().GetHubs()))
```

Comment:
Hi @ethervoid , the attempt to make this thread safe didn't quite work.  The check for GetHubs() == 0 above may change by the time we get to this point.  It happened for me running unit tests in CI:
```
panic: runtime error: integer divide by zero

goroutine 617565 [running]:
github.com/mattermost/mattermost-server/v5/app.(*App).GetHubForUserId(0xc045aa0d20, 0xc0041df0e0, 0x1a, 0xbf9db5ec1e284c9f)
	/mattermost/mattermost-server/app/web_hub.go:147 +0x2ca
github.com/mattermost/mattermost-server/v5/app.(*App).UpdateWebConnUserActivity(0xc045aa0d20, 0xc05ae70420, 0x1a, 0xc05ae70440, 0x1a, 0x1717b67bd76, 0x0, 0x1717b67bd76, 0xc0041df0e0, 0x1a, ...)
	/mattermost/mattermost-server/app/web_hub.go:331 +0x4f
github.com/mattermost/mattermost-server/v5/app.(*App).UpdateLastActivityAtIfNeeded(0xc045aa0d20, 0xc05ae70420, 0x1a, 0xc05ae70440, 0x1a, 0x1717b67bd76, 0x0, 0x1717b67bd76, 0xc0041df0e0, 0x1a, ...)
	/mattermost/mattermost-server/app/session.go:273 +0xea
github.com/mattermost/mattermost-server/v5/app.(*App).NewWebConn.func1()
	/mattermost/mattermost-server/app/web_conn.go:52 +0xc4
github.com/mattermost/mattermost-server/v5/app.(*Server).Go.func1(0xc003e88540, 0xc040a75000)
	/mattermost/mattermost-server/app/server.go:463 +0x2d
created by github.com/mattermost/mattermost-server/v5/app.(*Server).Go
	/mattermost/mattermost-server/app/server.go:462 +0x59
```

Reply:
Nice catch. I don't think this _resulted_ from adding the mutex. It would have been there even before fixing the race. Sending a PR.

Reply:
Hmm .. actually I need to go deeper to understand how this is happening. Can you point out to me which test failed ?

The hub count can only be set to zero from `HubStop` method. And this crash is happening from `NewWebConn` which only gets called from api4/websocket.go. But the stack trace shows it is being called from inside the `(*Server).Go` method. I don't see how that is happening. So somehow we are calling `NewWebConn` _from_ a goroutine, _after_ stopping the hub ?

Reply:
Until I can take a look I'll bet that the carrier is something done in a test but ill take a look as well. Thank you for pointing this out @wiggin77 :)

Reply:
@wiggin77 - do you remember which test was this ? I think it might be related to an investigation I am doing.

Reply:
@agnivade , here is the failed CI report:  https://circleci.com/gh/mattermost/mattermost-server/40105

Reply:
Thanks. I see it now. This needs some more thinking to fix it properly.

Reply:
Running that test using the `-race` we get a non-related race condition:

```
$ gotestsum --format testname -- -count=1 -race -mod=vendor -v -timeout=2000s -covermode=atomic -coverpkg=github.com/mattermost/mattermost-server/v5/app -run="TestHubStopRaceCondition" ./app 
Created temporary postgres database dbf1azzrk1p78ymx48ufbyxdqzro
=== RUN   TestHubStopRaceCondition
==================
WARNING: DATA RACE
Write at 0x00c0021553d0 by goroutine 49:
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:169 +0x42
  io.(*multiWriter).Write()
      /usr/local/go/src/io/multi.go:60 +0xbb
  go.uber.org/zap/zapcore.(*writerWrapper).Write()
      <autogenerated>:1 +0x87
  go.uber.org/zap/zapcore.(*ioCore).Write()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/core.go:90 +0x1c3
  go.uber.org/zap/zapcore.(*CheckedEntry).Write()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/entry.go:216 +0x1e7
  go.uber.org/zap.(*Logger).Debug()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/logger.go:179 +0x95
  github.com/mattermost/mattermost-server/v5/mlog.(*Logger).Debug-fm()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/mlog/log.go:163 +0x8f
  github.com/mattermost/mattermost-server/v5/plugin.(*PluginHealthCheckJob).run()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/plugin/health_check.go:31 +0x83

Previous write at 0x00c0021553d0 by goroutine 85:
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:169 +0x42
  io.(*multiWriter).Write()
      /usr/local/go/src/io/multi.go:60 +0xbb
  go.uber.org/zap/zapcore.(*writerWrapper).Write()
      <autogenerated>:1 +0x87
  go.uber.org/zap/zapcore.(*ioCore).Write()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/core.go:90 +0x1c3
  go.uber.org/zap/zapcore.(*CheckedEntry).Write()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/entry.go:216 +0x1e7
  go.uber.org/zap.(*Logger).Info()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/logger.go:187 +0x95
  github.com/mattermost/mattermost-server/v5/mlog.(*Logger).Info-fm()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/mlog/log.go:167 +0x8f
  github.com/mattermost/mattermost-server/v5/app.NewServer()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/server.go:256 +0x1793
  github.com/mattermost/mattermost-server/v5/app.setupTestHelper()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/helper_test.go:71 +0x621
  github.com/mattermost/mattermost-server/v5/app.Setup()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/helper_test.go:146 +0x183
  github.com/mattermost/mattermost-server/v5/app.TestHubStopRaceCondition()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_hub_test.go:71 +0x5f
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 49 (running) created at:
  github.com/mattermost/mattermost-server/v5/plugin.(*Environment).InitPluginHealthCheckJob()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/plugin/environment.go:510 +0x408
  github.com/mattermost/mattermost-server/v5/app.NewServer()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/server.go:246 +0x351b
  github.com/mattermost/mattermost-server/v5/app.setupTestHelper()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/helper_test.go:71 +0x621
  github.com/mattermost/mattermost-server/v5/app.Setup()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/helper_test.go:146 +0x183
  github.com/mattermost/mattermost-server/v5/app.TestHubStopRaceCondition()
      /home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_hub_test.go:71 +0x5f
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
```

Reply:
Running 1000 tests (count=1000) I've got this panic related to the data race above mentioned:

```
=== RUN   TestHubStopRaceCondition
panic: Log in goroutine after TestHubStopRaceCondition has completed

goroutine 25247 [running]:
testing.(*common).logDepth(0xc0e5c4c600, 0xc0e0624a80, 0xba, 0x3)
	/usr/local/go/src/testing/testing.go:678 +0x406
testing.(*common).log(...)
	/usr/local/go/src/testing/testing.go:658
testing.(*common).Log(0xc0e5c4c600, 0xc0b1561460, 0x1, 0x1)
	/usr/local/go/src/testing/testing.go:686 +0x6a
github.com/mattermost/mattermost-server/v5/mlog.(*testingWriter).Write(0xc0a37fdda0, 0xc0c38d0400, 0xba, 0x400, 0x818a29, 0xc0013f7cc0, 0x400)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/mlog/testing.go:21 +0xe4
io.(*multiWriter).Write(0xc0488ff160, 0xc0c38d0400, 0xba, 0x400, 0x4e, 0x111, 0xc0c38d0400)
	/usr/local/go/src/io/multi.go:60 +0x87
go.uber.org/zap/zapcore.(*ioCore).Write(0xc0c01682a0, 0x2, 0xbfa426bc24802125, 0xef4fc0a8b, 0x29104a0, 0x0, 0x0, 0x1923df1, 0x10, 0x1, ...)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/core.go:90 +0x106
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc020110420, 0xc0b7b6a8c0, 0x1, 0x1)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/entry.go:216 +0x117
go.uber.org/zap.(*Logger).Error(0xc0c89a3f80, 0x1923df1, 0x10, 0xc0b7b6a8c0, 0x1, 0x1)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/logger.go:203 +0x7f
github.com/mattermost/mattermost-server/v5/mlog.(*Logger).Error(...)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/mlog/log.go:175
github.com/mattermost/mattermost-server/v5/app.(*WebConn).IsAuthenticated(0xc000230dc0, 0xbfa426d5246dafc8)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:273 +0x24e
github.com/mattermost/mattermost-server/v5/app.(*WebConn).readPump.func2(0x0, 0x0, 0x0, 0x0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:144 +0x9e
github.com/gorilla/websocket.(*Conn).advanceFrame(0xc00c60eb00, 0x18, 0xc004dc2e38, 0xc0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/websocket/conn.go:925 +0x64e
github.com/gorilla/websocket.(*Conn).NextReader(0xc00c60eb00, 0xc004db4c00, 0xbfa426c62328cb23, 0x1843d44482, 0xc004dc2f00, 0x40e758)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/websocket/conn.go:980 +0xa0
github.com/gorilla/websocket.(*Conn).ReadJSON(0xc00c60eb00, 0x16eafc0, 0xc0022a25b0, 0x29104a0, 0x0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/websocket/json.go:50 +0x2f
github.com/mattermost/mattermost-server/v5/app.(*WebConn).readPump(0xc000230dc0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:154 +0x1ec
github.com/mattermost/mattermost-server/v5/app.(*WebConn).Pump(0xc000230dc0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:129 +0xa7
created by github.com/mattermost/mattermost-server/v5/app.registerDummyWebConn
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_hub_test.go:48 +0x2a2
FAIL	github.com/mattermost/mattermost-server/v5/app	64.373s
FAIL app.TestHubStopRaceCondition (64.37s)

=== Failed
=== FAIL: app TestHubStopRaceCondition (64.37s)
panic: Log in goroutine after TestHubStopRaceCondition has completed

goroutine 25247 [running]:
testing.(*common).logDepth(0xc0e5c4c600, 0xc0e0624a80, 0xba, 0x3)
	/usr/local/go/src/testing/testing.go:678 +0x406
testing.(*common).log(...)
	/usr/local/go/src/testing/testing.go:658
testing.(*common).Log(0xc0e5c4c600, 0xc0b1561460, 0x1, 0x1)
	/usr/local/go/src/testing/testing.go:686 +0x6a
github.com/mattermost/mattermost-server/v5/mlog.(*testingWriter).Write(0xc0a37fdda0, 0xc0c38d0400, 0xba, 0x400, 0x818a29, 0xc0013f7cc0, 0x400)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/mlog/testing.go:21 +0xe4
io.(*multiWriter).Write(0xc0488ff160, 0xc0c38d0400, 0xba, 0x400, 0x4e, 0x111, 0xc0c38d0400)
	/usr/local/go/src/io/multi.go:60 +0x87
go.uber.org/zap/zapcore.(*ioCore).Write(0xc0c01682a0, 0x2, 0xbfa426bc24802125, 0xef4fc0a8b, 0x29104a0, 0x0, 0x0, 0x1923df1, 0x10, 0x1, ...)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/core.go:90 +0x106
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc020110420, 0xc0b7b6a8c0, 0x1, 0x1)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/zapcore/entry.go:216 +0x117
go.uber.org/zap.(*Logger).Error(0xc0c89a3f80, 0x1923df1, 0x10, 0xc0b7b6a8c0, 0x1, 0x1)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/go.uber.org/zap/logger.go:203 +0x7f
github.com/mattermost/mattermost-server/v5/mlog.(*Logger).Error(...)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/mlog/log.go:175
github.com/mattermost/mattermost-server/v5/app.(*WebConn).IsAuthenticated(0xc000230dc0, 0xbfa426d5246dafc8)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:273 +0x24e
github.com/mattermost/mattermost-server/v5/app.(*WebConn).readPump.func2(0x0, 0x0, 0x0, 0x0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:144 +0x9e
github.com/gorilla/websocket.(*Conn).advanceFrame(0xc00c60eb00, 0x18, 0xc004dc2e38, 0xc0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/websocket/conn.go:925 +0x64e
github.com/gorilla/websocket.(*Conn).NextReader(0xc00c60eb00, 0xc004db4c00, 0xbfa426c62328cb23, 0x1843d44482, 0xc004dc2f00, 0x40e758)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/websocket/conn.go:980 +0xa0
github.com/gorilla/websocket.(*Conn).ReadJSON(0xc00c60eb00, 0x16eafc0, 0xc0022a25b0, 0x29104a0, 0x0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/websocket/json.go:50 +0x2f
github.com/mattermost/mattermost-server/v5/app.(*WebConn).readPump(0xc000230dc0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:154 +0x1ec
github.com/mattermost/mattermost-server/v5/app.(*WebConn).Pump(0xc000230dc0)
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_conn.go:129 +0xa7
created by github.com/mattermost/mattermost-server/v5/app.registerDummyWebConn
	/home/ethervoid/go/src/github.com/mattermost/mattermost-server/app/web_hub_test.go:48 +0x2a2
FAIL	github.com/mattermost/mattermost-server/v5/app	64.373s
```

Taking a look at the code we have a possible data race between we check [here](https://github.com/mattermost/mattermost-server/blob/master/app/web_hub.go#L146) the number of hubs and [here](https://github.com/mattermost/mattermost-server/blob/master/app/web_hub.go#L152) where we use them.

The only possible case is that we call the `HubStop` method, after that we create or we were creating additional hubs, and enter in that routine asking for the id while the stop methods are happening.

That situation is highly improbable and is only happening in the test but we should take a look and see if we can fix that part

Reply:
Yes, would you mind filing a ticket for this ? I want to improve this hub lock to reduce the lock contention.

Reply:
Created https://mattermost.atlassian.net/browse/MM-24792